applications/dashboard/models/class.permissionmodel.php

```diff
@@ -820,6 +965,47 @@ public function SaveAll($Permissions, $AllWhere = NULL) {
    }
 
    /**
+    * Reset permissions for all roles, based on the value in their Type column.
+    */
+   public static function ResetAllRoles() {
+      // Retrieve an array containing all available roles.
+      $RoleModel = new RoleModel();
+      $AllRoles = $RoleModel->GetArray();
+
+      // Iterate through our roles and reset their permissions.
+      $Permissions = Gdn::PermissionModel();
+      foreach ($AllRoles as $RoleID => $Role) {
+         $Permissions->ResetRole($Role);
+      }
+   }
+
+   /**
+    * Reset permissions for a role, based on the value in its Type column.
+    *
+    * @param string $Role Name of the role to reset permissions for.
+    * @throws Exception
+    */
+   public function ResetRole($Role) {
+      // Grab the value of Type for this role.
+      $RoleType = $this->SQL->GetWhere('Role', array('Name' => $Role))->Value('Type');
+
+      if ($RoleType == '') {
+         $RoleType = self::TYPE_MEMBER;
+      }
+
+      $this->AssignDefaults();
```

Comment:
You have an inefficiency here where calling `ResetAllRoles()` means that this is called multiple times even though one call gives enough information for all defaults.

I'd suggest making an accessor for `$_DefaultPermissions` and have that lazy populate the array. Also FYI: we don't use underscores on protected properties or methods anymore. That was an artifact of old PHP.